File: 04.CLIESC/CLIESC_Comercializadora_RESTJAVA_GR05/src/main/java/ec/edu/monster/services/CarritoLocalService.java
package ec.edu.monster.services;

import ec.edu.monster.models.CarritoItem;
import ec.edu.monster.models.DetalleFacturaRequest;
import ec.edu.monster.models.SolicitudFactura;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Servicio Singleton para gestionar el carrito de compras en memoria local
 */
public class CarritoLocalService {

    private static CarritoLocalService instance;
    private Map<String, CarritoItem> items; // Key: codigo del producto

    private CarritoLocalService() {
        this.items = new HashMap<>();
    }

    /**
     * Obtiene la instancia única del servicio
     */
    public static synchronized CarritoLocalService getInstance() {
        if (instance == null) {
            instance = new CarritoLocalService();
        }
        return instance;
    }

    /**
     * Agrega un producto al carrito o incrementa la cantidad si ya existe
     */
    public void agregarProducto(CarritoItem item) {
        if (item == null || item.getCodigo() == null) {
            throw new IllegalArgumentException("El item y su código no pueden ser nulos");
        }

        String codigo = item.getCodigo();

        if (items.containsKey(codigo)) {
            // Si ya existe, incrementar la cantidad
            CarritoItem existente = items.get(codigo);
            existente.setCantidad(existente.getCantidad() + item.getCantidad());
        } else {
            // Si no existe, agregar nuevo item
            items.put(codigo, item);
        }
    }

    /**
     * Remueve un producto del carrito por su código
     */
    public void removerProducto(String codigo) {
        items.remove(codigo);
    }

    /**
     * Obtiene todos los items del carrito
     */
    public List<CarritoItem> obtenerItems() {
        return new ArrayList<>(items.values());
    }

    /**
     * Obtiene la cantidad total de items (diferentes productos)
     */
    public int obtenerCantidadItems() {
        return items.size();
    }

    /**
     * Calcula el precio total del carrito
     */
    public BigDecimal obtenerTotal() {
        return items.values().stream()
                .map(CarritoItem::getSubtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    /**
     * Vacía completamente el carrito
     */
    public void vaciarCarrito() {
        items.clear();
    }

    /**
     * Verifica si el carrito está vacío
     */
    public boolean estaVacio() {
        return items.isEmpty();
    }

    /**
     * Convierte el carrito actual a una SolicitudFactura para enviar al servidor
     */
    public SolicitudFactura convertirASolicitudFactura(String cedula, String tipoPago, Integer numeroCuotas) {
        if (estaVacio()) {
            throw new IllegalStateException("El carrito está vacío");
        }

        SolicitudFactura solicitud = new SolicitudFactura();
        solicitud.setCedula(cedula);
        solicitud.setTipoPago(tipoPago);
        solicitud.setNumeroCuotas(numeroCuotas);

        List<DetalleFacturaRequest> detalles = new ArrayList<>();
        for (CarritoItem item : items.values()) {
            DetalleFacturaRequest detalle = new DetalleFacturaRequest();
            detalle.setCodigo(item.getCodigo());
            detalle.setCantidad(item.getCantidad());
            detalles.add(detalle);
        }
        solicitud.setDetalles(detalles);

        return solicitud;
    }
}

File: 04.CLI MOV/CLIMOV_ComerElect_RESTJAVA_GR05/app/src/main/java/ec/edu/monster/controlador/carrito/CarritoLocalService.kt
package ec.edu.monster.controlador.carrito

import android.util.Log
import ec.edu.monster.modelo.ElectrodomesticoDTO
import java.math.BigDecimal

/**
 * Servicio de carrito LOCAL (no hace llamadas al servidor).
 * El carrito se guarda en memoria y se usa para crear facturas.
 */
class CarritoLocalService private constructor() {
    
    data class ItemCarritoLocal(
        val producto: ElectrodomesticoDTO,
        var cantidad: Int
    ) {
        val subtotal: BigDecimal
            get() = (producto.precio ?: BigDecimal.ZERO).multiply(BigDecimal(cantidad))
    }
    
    private val items = mutableMapOf<String, ItemCarritoLocal>() // codigo -> item
    
    companion object {
        @Volatile
        private var INSTANCE: CarritoLocalService? = null
        
        fun getInstance(): CarritoLocalService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: CarritoLocalService().also { INSTANCE = it }
            }
        }
    }
    
    /**
     * Agrega un producto al carrito o incrementa su cantidad si ya existe
     */
    fun agregar(producto: ElectrodomesticoDTO, cantidad: Int) {
        Log.d("CARRITO_LOCAL", "Agregando: ${producto.nombre} x$cantidad")
        
        val item = items[producto.codigo]
        if (item != null) {
            // Ya existe, incrementar cantidad
            item.cantidad += cantidad
            Log.d("CARRITO_LOCAL", "Producto ya existía. Nueva cantidad: ${item.cantidad}")
        } else {
            // Nuevo producto
            items[producto.codigo] = ItemCarritoLocal(producto, cantidad)
            Log.d("CARRITO_LOCAL", "Producto agregado como nuevo")
        }
        
        Log.d("CARRITO_LOCAL", "Total items en carrito: ${items.size}")
    }
    
    /**
     * Obtiene todos los items del carrito
     */
    fun obtenerItems(): List<ItemCarritoLocal> {
        Log.d("CARRITO_LOCAL", "Obteniendo items del carrito: ${items.size} productos")
        return items.values.toList()
    }
    
    /**
     * Remueve un producto del carrito
     */
    fun remover(codigo: String) {
        Log.d("CARRITO_LOCAL", "Removiendo producto: $codigo")
        items.remove(codigo)
        Log.d("CARRITO_LOCAL", "Items restantes: ${items.size}")
    }
    
    /**
     * Actualiza la cantidad de un producto
     */
    fun actualizarCantidad(codigo: String, nuevaCantidad: Int) {
        Log.d("CARRITO_LOCAL", "Actualizando cantidad de $codigo a $nuevaCantidad")
        val item = items[codigo]
        if (item != null) {
            item.cantidad = nuevaCantidad
            Log.d("CARRITO_LOCAL", "Cantidad actualizada")
        } else {
            Log.w("CARRITO_LOCAL", "Producto no encontrado en carrito")
        }
    }
    
    /**
     * Limpia todo el carrito
     */
    fun limpiar() {
        Log.d("CARRITO_LOCAL", "Limpiando carrito completo")
        items.clear()
    }
    
    /**
     * Calcula el total del carrito
     */
    fun calcularTotal(): BigDecimal {
        return items.values.sumOf { it.subtotal }
    }
    
    /**
     * Obtiene la cantidad de items distintos en el carrito
     */
    fun cantidadItems(): Int {
        return items.size
    }
}
